crates/subspace-farmer/src/farmer_cache/piece_cache_state.rs

```diff
@@ -141,12 +143,58 @@ where
         dangling_free_offsets.clear();
         (stored_pieces, dangling_free_offsets)
     }
+
+    pub(super) fn should_replace(
+        &mut self,
+        key: &KeyWithDistance,
+    ) -> Option<(KeyWithDistance, FarmerCacheOffset<CacheIndex>)> {
+        if !self.should_include_key_internal(key) {
+            return None;
+        }
+
+        if !self.has_free_capacity() {
+            self.stored_pieces.pop_last()
+        } else {
+            None
+        }
+    }
+
+    pub(super) fn should_include_key(&self, peer_id: PeerId, key: PieceIndex) -> bool {
+        let key = KeyWithDistance::new(peer_id, key.to_multihash());
+        self.should_include_key_internal(&key)
+    }
+
+    fn has_free_capacity(&self) -> bool {
+        if !self.dangling_free_offsets.is_empty() {
+            return true;
+        }
+
+        self.backends.iter().any(|backend| backend.free_size() > 0)
+    }
+
+    fn should_include_key_internal(&self, key: &KeyWithDistance) -> bool {
+        if self.stored_pieces.contains_key(key) {
+            return false;
+        }
+
+        if self.has_free_capacity() {
+            return true;
+        }
+
+        let top_key = self.stored_pieces.last_key_value().map(|(key, _)| key);
+
+        if let Some(top_key) = top_key {
+            top_key > key
+        } else {
+            false // TODO: consider adding error here
```

Comment:
What error is this? Looks like it needs to be initialized with empty list of backends for this to happen, which is possible and not an error.